### DISCUSSION TITLE
ABACUS Container Solution

### DISCUSSION THREAD

**Discussion by tansongchen** [deepmodeling/abacus-develop#87] [Ideas]
The following notes are extracted from discussions with @darelbeida .

# Requirements

1. Container should be able to use in both development and production
2. Container should support both Intel toolkits and GNU toolkits
3. Container should be defined by a Dockerfile, be uploaded to Deep Modeling registry by core developers and be able to be downloaded by everyone

# Technical Background

- Docker supports multi-stage build and build time arguments
- GitHub restricts access of repository secrets when a pull request is raised from a fork

# Solution

1. There exists a `Dockerfile` file at the root directory of this repository, which
    1. By accepting `--target` option, can build to either the "development kit" container (with only dependencies) and the "production" container (with only the compiled binary)
    2. By accepting `--build-args` option, can build with either Intel toolkits and GNU toolkits
2. There exists an action script `test.yml` that is invoked upon pull requests toward "develop" branch, building the `Dockerfile` from the head branch (with `--cache-from`) targeting the "development kit" container, mounting the code from the head branch, and running CMake & CTest inside the container
3. There exists an action script `container.yml` that is invoked after merges happened on "develop" branch, building the `Dockerfile` from "develop" branch targeting both the "development kit" container and "production" container, uploading them to Deep Modeling registry (https://ghcr.io/deepmodeling/)

# Action

#87

**Comment by caic99** [comment]
Solved in #353
